packages-ext/telemetry/src/utils/generate-text.ts

```diff
@@ -0,0 +1,143 @@
+import { getTracer } from './get-tracer'
+import { recordSpan } from './record-span'
+import { stringifyTool } from './stringify-tool'
+import { wrapTool } from './wrap-tool'
+
+export const generateText = async (options: GenerateTextOptions) => {
+  const tracer = getTracer()
+
+  const commonAttributes = (operationId: string) => ({
+    'ai.model.id': options.model,
+    // TODO: provider name
+    'ai.model.provider': 'xsai',
+    'ai.operationId': operationId,
+    'ai.response.providerMetadata': '{}',
+    'operation.name': operationId,
+  })
+
+  const idAttributes = () => {
+    const id = crypto.randomUUID()
+
+    return {
+      'ai.response.id': id,
+      'ai.response.timestamp': new Date().toISOString(),
+      'gen_ai.response.id': id,
+    }
+  }
+
+  const rawGenerateText = async (options: GenerateTextOptions): Promise<TrampolineFn<GenerateTextResult>> => {
+    const messages: Message[] = structuredClone(options.messages)
+    const steps: CompletionStep<true>[] = options.steps ? structuredClone(options.steps) : []
+
+    const [stepWithoutToolCalls, { messages: msgs1, msgToolCalls, reasoningText }] = await recordSpan({
+      attributes: {
+        ...commonAttributes('ai.generateText.doGenerate'),
+        ...idAttributes(),
+        ...(options.tools != null && options.tools.length > 0
+          ? {
+              'ai.prompt.toolChoice': JSON.stringify(options.toolChoice ?? { type: 'auto' }),
+              'ai.prompt.tools': options.tools.map(stringifyTool),
+            }
+          : {}),
+        'ai.prompt.messages': JSON.stringify(messages),
+        'ai.response.model': options.model,
+        'gen_ai.request.model': options.model,
+        'gen_ai.response.id': crypto.randomUUID(),
+        'gen_ai.response.model': options.model,
+        'gen_ai.system': 'xsai',
+      },
+      name: 'ai.generateText.doGenerate',
+      tracer,
+    }, async (span) => {
+      const res = await chat({
+        ...options,
+        maxSteps: undefined,
+        steps: undefined,
+        stream: false,
+      })
+        .then(responseJSON<GenerateTextResponse>)
+
+      const [step, { messages: msgs, msgToolCalls, reasoningText }] = await extractGenerateTextStep({
+        ...options,
+        messages,
+        steps,
+      }, res)
+
+      span.setAttributes({
+        ...((step.text != null && step.toolCalls.length === 0) ? { 'ai.response.text': step.text } : {}),
+        ...(step.toolCalls.length > 0 ? { 'ai.response.toolCalls': JSON.stringify(step.toolCalls) } : {}),
+        'ai.response.finishReason': step.finishReason,
+        'ai.usage.completionTokens': step.usage.completion_tokens,
+        'ai.usage.promptTokens': step.usage.prompt_tokens,
+        'gen_ai.response.finish_reasons': [step.finishReason],
+        'gen_ai.usage.input_tokens': step.usage.prompt_tokens,
+        'gen_ai.usage.output_tokens': step.usage.completion_tokens,
+      })
```

Comment:
Be aware that this covers trace only, for metrics, we need to perform increment call to metric counter. Maybe add // TODO mark here for me to implement metrics later.